crates/bevy_ecs/src/query/iter.rs

```diff
@@ -39,6 +39,146 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> QueryIter<'w, 's, Q, F> {
             cursor: QueryIterationCursor::init(world, query_state, last_change_tick, change_tick),
         }
     }
+
+    /// Executes the equivalent of [`Iterator::fold`] over a contiguous segment
+    /// from an archetype.
+    ///
+    /// # Safety
+    ///  - all `rows` must be in `[0, tables.entity_count)`.
+    ///  - `table` must match Q and F
+    ///  - Both `Q::IS_DENSE` and `F::IS_DENSE` must be true.
+    #[inline]
+    pub(super) unsafe fn fold_table<B, Func>(
+        &mut self,
+        mut accum: B,
+        func: &mut Func,
+        table: &'w Table,
+        rows: Range<usize>,
+    ) -> B
+    where
+        Func: FnMut(B, Q::Item<'w>) -> B,
+    {
+        Q::set_table(&mut self.cursor.fetch, &self.query_state.fetch_state, table);
+        F::set_table(
+            &mut self.cursor.filter,
+            &self.query_state.filter_state,
+            table,
+        );
+
+        let entities = table.entities();
+        for row in rows {
+            // SAFETY: set_table was called prior.
+            // `current_row` is a table row in range of the current table, because if it was not, then the if above would have been executed.
+            let entity = entities.get_unchecked(row);
+            let row = TableRow::new(row);
+            if !F::filter_fetch(&mut self.cursor.filter, *entity, row) {
```

Comment:
Style nit, but I really don't like condensing safety comments in this way. IMO it's a lot clearer if each unsafe operation has its own comment, directly above it. The corresponding code in `fold_archetype` has this same issue as well.

Reply:
Split the safety comment.